docs/internals/data_file.md

```diff
@@ -0,0 +1,182 @@
+superblock on disk (so the logical state reachable from the superblock stays the same). Only after a
+relatively large amount of new grid blocks written, replica atomically writes the new superblock,
+with a new free set and a new logical "root pointer". To implement atomic update of the superblock,
+the superblock is physically stored as 4 distinct copies on disk. After startup, replica picks the
+latest superblock which has at least 2 copies written. Picking just the latest copy would be wrong
+--- unlike the grid blocks, the superblock stores its own checksum, and is vulnerable to misdirected
+reads (i.e., a misdirected read can hide the sole latest copy).
+
+Because the superblock (and hence, logical grid state) is updated infrequently and in bursts, it
+can't represent the entirety of persistent state. The rest of the state is stored in the write ahead
+log. WAL is a ring buffer with prepares, and represents the logical diff which should be applied to
+the state represented by superblock/grid to get the actual current state of the system. When a
+replica processes a prepare it, roughly:
+
+* writes it to WAL on disk
+* applies changes from prepare to the in-memory data structure representing the current state
+* applies changes from the prepare to the pending on-disk state by allocating and writing fresh grid
+  blocks
+
+When enough prepares are received, the superblock is updated to point the accumulated-so-far new
+disk state.
+
+This covers how the three major areas of the data file -- the write-ahead log, the superblock and
+the grid -- work together to represent abstract persistent logical state.
+
+Concretely, the sate of TigerBeetle is a collection (forest) of LSM trees. LSM structure is
+described [in a separate document](./lsm.md), here only high level on-disk layout is discussed.
+
+Each LSM tree stores a set of values. Values are:
+
+* fixed in size,
+* small (hundreds of bytes),
+* sorted by key,
+* which is embedded in the value itself (e.g, an `Account` value uses `timestamp` as a unique key).
+
+To start from the middle, values are arranged in tables on disk. Each table represents a sorted
+array of values and is physically stored in multiple blocks. Specifically:
+
+* table's data blocks are just sorted array of objects
+* table's index block stores pointers to the data blocks, as well as boundary keys.
+
+```zig
+const TableDataBlock = struct {
+    values_sorted: [value_count_max]Value,
+};
+
+const TableIndexBlock = struct {
+    data_block_checksums: [data_block_count_max]u128,
+    data_block_indexes:   [data_block_count_max]u64,
+    data_block_key_max:   [data_block_count_max]Key,
+};
+
+const TableInfo = struct {
+    tree_id: u16,
+    index_block_index: u64,
+    index_block_checksum: u128,
+    key_min: Key,
+    key_max: Key,
+};
+```
+
+To lookup an object in a table, binary search the index block to locate the data block which should
+hold the object, then binary search inside the data block.
```

Comment:
```suggestion
To lookup a value in a table, binary search the index block to locate the data block which should
hold the value, then binary search inside the data block.
```

Typically we use "object" to refer to e.g. account/transfer, whereas "value" is broader and can include secondary index "(key, timestamp)" tuples too.